include/fiction/algorithms/simulation/sidb/defect_operational_domain.hpp

```diff
@@ -0,0 +1,812 @@
+        {
+            if (layout.is_empty_cell({decr_x, incr_y}))
+            {
+                neighbors.emplace_back(decr_x, incr_y);
+            }
+        }
+        // up
+        if (moore.y != incr_y)
+        {
+            if (layout.is_empty_cell({moore.x, incr_y}))
+            {
+                neighbors.emplace_back(moore.x, incr_y);
+            }
+        }
+        // upper-right
+        if (moore.x != incr_x && moore.y != incr_y)
+        {
+            if (layout.is_empty_cell({incr_x, incr_y}))
+            {
+                neighbors.emplace_back(incr_x, incr_y);
+            }
+        }
+
+        return neighbors;
+    };
+    /**
+     * The SiDB cell-level layout to investigate.
+     */
+    sidb_defect_surface<Lyt> layout{};
+    /**
+     * The specification of the layout.
+     */
+    const std::vector<TT>& truth_table;
+    /**
+     * The parameters for the operational domain computation.
+     */
+    const defect_operational_domain_params& params;
+    /**
+     * North-west cell.
+     */
+    typename Lyt::cell nw_cell{};
+    /**
+     * The north-west bounding box of the layout.
+     */
+    typename Lyt::cell nw_bb_layout{};
+    /**
+     * South-east cell.
+     */
+    typename Lyt::cell se_cell{};
+    /**
+     * The south-east bounding box of the layout.
+     */
+    typename Lyt::cell se_bb_layout{};
+    /**
+     * The current defect position.
+     */
+    typename Lyt::cell current_defect_position{};
+    /**
+     * The previous defect position.
+     */
+    typename Lyt::cell previous_defect_position{};
+    /**
+     * The operational domain of the layout.
+     */
+    defect_operational_domain<Lyt> defect_op_domain{};
+    /**
+     * The statistics of the operational domain computation.
+     */
+    defect_operational_domain_stats& stats;
+    /**
+     * Number of simulator invocations.
+     */
+    std::atomic<std::size_t> num_simulator_invocations{0};
+    /**
+     * Number of evaluated defect positions.
+     */
+    std::atomic<std::size_t> num_evaluated_defect_positions{0};
+};
+
+}  // namespace detail
+
+/**
+ * Computes the defect operational domain of the given SiDB layout. The defect operational domain is the set of all
+ * defect positions for which the layout is logically operational. Logical operation is defined as the layout
+ * implementing the given truth table.
+ *
+ * This algorithm uses a grid search to determine the defect operational domain. The grid search is performed
+ * by exhaustively sweeping all possible atomic defect positions in x and y dimensions.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @tparam TT Truth table type.
+ * @param lyt Layout to compute the defect operational domain for.
+ * @param spec Expected Boolean function of the layout given as a multi-output truth table.
+ * @param step_size The parameter specifying the interval between consecutive defect positions to be evaluated.
```

Comment:
Why is this not a parameter for QuickCell?

Reply:
you mean _QuickTrace_, right? 

I think in _QuickTrace_ we cannot skip defect positions. We might lose the contour when skipping positions.

Reply:
Yes, sorry :man_facepalming: The Quick-... family of algorithms started to confuse me :sweat_smile:

Reply:
Sorry, yeah, the family grew very fast 🧑‍🧑‍🧒‍🧒